lib/internal/Magento/Framework/DataObject/Copy.php

```diff
@@ -61,28 +61,240 @@ public function __construct(
      */
     public function copyFieldsetToTarget($fieldset, $aspect, $source, $target, $root = 'global')
     {
-        if (!$this->_isFieldsetInputValid($source, $target)) {
+        $sourceIsArray = is_array($source);
```

Comment:
This method becomes much more complex than it should be. There is duplication of logic and the complexity of the method increases.

You can do following:
1. wrap the information about isDataObject, isExtensible, etc in some simple class, i.e. and pass this object as input to the methods, where currently object is passed.
```
class DataObjectTypeWrapper {
    public function __construct($object) {}

    public function getObject() {}

    public function isDataObject() {}

    ...
}
```
2. Move every "if" body to a separate method (and potentially separate class) which accepts the DataOjectTypeWrapper. There are a lot of duplication of logic right now which an be avoided by re-suing those methods.